sys-libs/zlib: add patch to fix configure AR/RANLIB/NM detection

10ne1 opened the pull request:
This adds a patch from the zlib-devel ML to fix a minor
build inconsistency leading to GNU tools usage in LLVM
configured builds.

The reason we're adding it directly here is zlib upstream
takes forever to release new versions and there's not much
activity for some time (last release was in Jan 2017).

Bug: https://bugs.gentoo.org/831628
Signed-off-by: Adrian Ratiu <adrian.ratiu@collabora.com>

thesamesam commented on sys-libs/zlib/files/zlib-1.2.11-configure-fix-AR-RANLIB-NM-detection.patch:
```diff
@@ -0,0 +1,74 @@
+From 6221a72528b30e05941309e49808da3ef28da115 Mon Sep 17 00:00:00 2001
```
Please include a link to relevant Gentoo bugs and upstream references.

here, I'd include two lines at the top: one for the mailing list submission and one for the Gentoo bug.

10ne1 replied:
Done, Thanks!

10ne1 commented:
Hi @thesamesam Gentle ping, is this ready to be merged? Thank you!

thesamesam commented:
@10ne1 Can you rebase and fix the issue in adb9a35721e67faab0c5a75f31e38e8791e2f43b? Thanks!

10ne1 commented:
> @10ne1 Can you rebase and fix the issue in [adb9a35](https://github.com/gentoo/gentoo/commit/adb9a35721e67faab0c5a75f31e38e8791e2f43b)? Thanks!

@thesamesam It was due to a typo I missed in the `unset NM` case, I've fixed it and rebased. Can we please re-land this? Thank you!

thesamesam commented:
Doesn't build for me.

```
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -fdiagnostics-color=always -frecord-gcc-switches     -ggdb3 -D_LARGEFILE64_SOURCE=1 -DHAVE_HIDDEN -I. -I/var/tmp/portage/sys-libs/zlib-1.2.11-r5/work/zlib-1.2.11/ -D_FILE_OFFSET_BITS=64 -c -o example64.o /var/tmp/portage/sys-libs/zlib-1.2.11-r5/work/zlib-1.2.11/test/example.c
x86_64-pc-linux-gnu-gcc -O2 -pipe -march=native -fdiagnostics-color=always -frecord-gcc-switches     -ggdb3 -D_LARGEFILE64_SOURCE=1 -DHAVE_HIDDEN -I. -I/var/tmp/portage/sys-libs/zlib-1.2.11-r5/work/zlib-1.2.11/ -D_FILE_OFFSET_BITS=64 -c -o minigzip64.o /var/tmp/portage/sys-libs/zlib-1.2.11-r5/work/zlib-1.2.11/test/minigzip.c
rc libz.a adler32.o crc32.o deflate.o infback.o inffast.o inflate.o inftrees.o trees.o zutil.o compress.o uncompr.o gzclose.o gzlib.o gzread.o gzwrite.o
make: rc: No such file or directory
make: *** [Makefile:131: libz.a] Error 127
make: *** Waiting for unfinished jobs....
 * ERROR: sys-libs/zlib-1.2.11-r5::gentoo failed (compile phase):
 *   emake failed
 *
```

For bits like this, if not already, suggest trying it in a Gentoo Docker image or chroot (or real system 😉) just to catch some of the standard cases which are a bit odd -- like how `AR` isn't actually set on Gentoo systems explicitly by default, we just have a sane fallback. I think here it's assuming AR is exported in the environment.

It should be an easy fix in the ebuild.

10ne1 commented:
That is weird, I tried this in a Gentoo docker image and it worked. I'll revisit, thanks for the feedback!

```
REPOSITORY            TAG       IMAGE ID       CREATED         SIZE
gentoo/stage3         latest    1957c7506a55   3 weeks ago     999MB
gentoo/portage        latest    9794a4a1de5b   3 weeks ago     242MB
```

ionenwks commented on sys-libs/zlib/files/zlib-1.2.11-configure-fix-AR-RANLIB-NM-detection.patch:
```diff
@@ -0,0 +1,79 @@
+-    test -n "${CROSS_PREFIX}" && echo Using ${RANLIB} | tee -a configure.log
+-else
+-    RANLIB=${RANLIB-"ranlib"}
++if [ -z "RANLIB" ] ; then
```
```suggestion
+if [ -z "$RANLIB" ] ; then
```

10ne1 replied:
Thanks! Done.

ionenwks commented on sys-libs/zlib/files/zlib-1.2.11-configure-fix-AR-RANLIB-NM-detection.patch:
```diff
@@ -0,0 +1,79 @@
+-else
+-    AR=${AR-"ar"}
+-    test -n "${CROSS_PREFIX}" && echo Using ${AR} | tee -a configure.log
++if [ -z "AR" ] ; then
```
```suggestion
+if [ -z "$AR" ] ; then
```

10ne1 replied:
Thanks! Done.

ionenwks commented on sys-libs/zlib/files/zlib-1.2.11-configure-fix-AR-RANLIB-NM-detection.patch:
```diff
@@ -0,0 +1,79 @@
+-else
+-    NM=${NM-"nm"}
++if [ -z "$NM" ] ; then
++  if "${CROSS_PREFIX}nm" --version >/dev/null 2>/dev/null || test $? -lt 126;   then
```
nitpick: original has a single space `126; then` but there's now several
(had a quick look to see if anything was out of place compared to original, fortunately this is all I found)

On another nitpicky note, the rest of the script uses `[ .. ]; then` while the new ones are `[ .. ] ; then` (extra space) and it's good to stay consistent. Not that style/spacing matter much unless this version is sent upstream.

10ne1 commented:
Thank you @ionenwks and @thesamesam I've applied the fixes and updated the patch to v3 which I've sent to zlib-devel upstream (still no activity unfortunately). I do not think we need to update the ebuild, just the patch fixes are enough. Sorry for the delay & churn.

For reference here is my Gentoo [docker image build log](https://gist.github.com/10ne1/e5f3eadec9e8a441daae4d2fc0ada91e).
